// @haiku/ssheldon118-artish2/code/main/code.js
var Haiku = require("@haiku/core");
module.exports = {
  metadata: {
    uuid: "6935558e-e9a6-43b6-a636-1e0f18f4f048",
    version: "0.0.3",
    organization: "ssheldon118",
    project: "artish2",
    branch: "master",
    folder: "/Users/stephensheldon/.haiku/projects/ssheldon118/artish2",
    username: "ssheldon118@gmail.com",
    root: "https://cdn.haiku.ai/0bf79054-8c65-4d79-8666-a0b321515252/a9c7061caa2358d4d96b594fb491a0f432fbf454/",
    core: "5.1.1",
    title: "Main",
    type: "haiku",
    relpath: "code/main/code.js"
  },
  options: {},
  states: {},
  eventHandlers: {},
  timelines: {
    Default: {
      "haiku:Main-03757d2ca1026e0a": {
        "style.WebkitTapHighlightColor": "rgba(0,0,0,0)",
        "style.position": "relative",
        "style.overflowX": "hidden",
        "style.overflowY": "hidden",
        "sizeAbsolute.x": 714.268,
        "sizeAbsolute.y": 448.233,
        "sizeMode.x": 1,
        "sizeMode.y": 1,
        "sizeMode.z": 1,
        opacity: 1
      },
      "haiku:Layer-1-Artish2-01-85cec093356d01e8": {
        "sizeMode.x": 1,
        x: "0px",
        viewBox: "0 0 777 481",
        enableBackground: "new 0 0 777 481",
        "xml:space": "preserve",
        "style.position": "absolute",
        "style.margin": "0",
        "style.padding": "0",
        "style.border": "0",
        y: "0px",
        "sizeMode.y": 1,
        "sizeAbsolute.x": 777,
        "sizeAbsolute.y": 481,
        "translation.x": 357.13349999999997,
        "translation.y": 224.11700000000002,
        "style.zIndex": 1,
        "scale.x": 0.919,
        "scale.y": 0.932,
        opacity: 1
      },
      "haiku:Rectangle-1-X-A0-Image-c6751370fcdb39b2": {
        overflow: "visible",
        width: 3237,
        height: 2000,
        "xlink:href":
          "web+haikuroot://assets/designs/artish2.ai.contents/artboards/artish2-01_image_1.png",
        "translation.y": 1,
        "scale.x": 0.24,
        "scale.y": 0.24
      },
      "haiku:Layer-1-Artish2-07-4f9f69f6a48787db": {
        "sizeMode.x": 1,
        x: "0px",
        viewBox: "0 0 776.88 480",
        enableBackground: "new 0 0 776.88 480",
        "xml:space": "preserve",
        "style.position": "absolute",
        "style.margin": "0",
        "style.padding": "0",
        "style.border": "0",
        y: "0px",
        "sizeMode.y": 1,
        "sizeAbsolute.x": 776.88,
        "sizeAbsolute.y": 480,
        "translation.x": {
          "0": { value: 272.578, edited: true },
          "83": { value: 272.578, curve: "easeInOutQuad", edited: true },
          "383": { value: 356.97636, edited: true }
        },
        "translation.y": 224.16000000000003,
        "style.zIndex": 3,
        "scale.x": 0.919,
        "scale.y": 0.934,
        opacity: {
          "0": { value: 0, edited: true },
          "150": { value: 0, curve: "linear", edited: true },
          "383": { value: 1, edited: true }
        }
      },
      "haiku:Layer-1-Artish2-04-d1c1411a17f9353a": {
        "sizeMode.x": 1,
        x: "0px",
        viewBox: "0 0 777 481",
        enableBackground: "new 0 0 777 481",
        "xml:space": "preserve",
        "style.position": "absolute",
        "style.margin": "0",
        "style.padding": "0",
        "style.border": "0",
        y: "0px",
        "sizeMode.y": 1,
        "sizeAbsolute.x": 777,
        "sizeAbsolute.y": 481,
        "translation.x": 357.4995,
        "translation.y": {
          "0": {
            value: 240.67049598220942,
            curve: "easeInOutQuad",
            edited: true
          },
          "267": { value: 224.08100609222504, edited: true }
        },
        "style.zIndex": 2,
        "scale.x": 0.919,
        "scale.y": {
          "0": { value: 1.001, curve: "easeInOutQuad", edited: true },
          "267": { value: 0.932, edited: true }
        },
        opacity: {
          "0": { value: 0, edited: true, curve: "easeInOutQuad" },
          "417": { value: 1, edited: true }
        }
      },
      "haiku:Layer-1-Artish2-03-f996e91f0f8c2e56": {
        "sizeMode.x": 1,
        x: "0px",
        viewBox: "0 0 777 481",
        enableBackground: "new 0 0 777 481",
        "xml:space": "preserve",
        "style.position": "absolute",
        "style.margin": "0",
        "style.padding": "0",
        "style.border": "0",
        y: "0px",
        "sizeMode.y": 1,
        "sizeAbsolute.x": 777,
        "sizeAbsolute.y": 481,
        "translation.x": {
          "0": { value: 391.439, edited: true },
          "150": { value: 391.439, edited: true, curve: "easeInOutQuad" },
          "433": { value: 357.1675102573147, edited: true }
        },
        "translation.y": {
          "0": { value: 241.05401611328128, edited: true },
          "150": {
            value: 237.66469370714287,
            edited: true,
            curve: "easeInOutQuad"
          },
          "433": { value: 223.92999914495087, edited: true }
        },
        "style.zIndex": 4,
        "scale.x": {
          "0": { value: 1.101, edited: true },
          "150": { value: 1.101, edited: true, curve: "easeInOutQuad" },
          "433": { value: 0.919, edited: true }
        },
        "scale.y": {
          "0": { value: 1.121, edited: true },
          "150": { value: 1.121, edited: true, curve: "easeInOutQuad" },
          "433": { value: 0.932, edited: true }
        },
        opacity: {
          "0": { value: 0, edited: true },
          "200": { value: 0, edited: true, curve: "linear" },
          "433": { value: 1, edited: true }
        }
      },
      "haiku:Ellipse-3-Copy-2-X-A0-Image-cdae8131d18a4358": {
        overflow: "visible",
        width: 438,
        height: 438,
        "xlink:href":
          "web+haikuroot://assets/designs/artish2.ai.contents/artboards/artish2-03_image_1.png",
        "translation.x": 82,
        "translation.y": 61,
        "scale.x": 0.24,
        "scale.y": 0.24
      },
      "haiku:Ellipse-2-Copy-3-X-A0-Image-4ecacd5172c17398": {
        overflow: "visible",
        width: 3237,
        height: 2000,
        "xlink:href":
          "web+haikuroot://assets/designs/artish2.ai.contents/artboards/artish2-04_image_1.png",
        "scale.x": 0.24,
        "scale.y": 0.24
      },
      "haiku:Ellipse-2-Copy-2-X-A0-Image-667e160c068a8e09": {
        overflow: "visible",
        width: 1975,
        height: 2000,
        "xlink:href":
          "web+haikuroot://assets/designs/artish2.ai.contents/artboards/artish2-07_image_1.png",
        "scale.x": 0.24,
        "scale.y": 0.24
      },
      "haiku:Asset-1-3dcb3812a3f867d9": {
        viewBox: "0 0 504.76 435.36",
        "style.position": "absolute",
        "style.margin": "0",
        "style.padding": "0",
        "style.border": "0",
        "sizeMode.x": 1,
        "sizeMode.y": 1,
        "sizeAbsolute.x": 504.76,
        "sizeAbsolute.y": 435.36,
        "translation.x": {
          "0": { value: 482.102, edited: true },
          "217": { value: 482.102, edited: true, curve: "easeInOutQuad" },
          "550": { value: 501.8859691012834, edited: true }
        },
        "translation.y": {
          "0": { value: 240.93751525878906, edited: true },
          "217": {
            value: 240.93751525878906,
            edited: true,
            curve: "easeInOutQuad"
          },
          "550": { value: 224.11586725173737, edited: true }
        },
        "style.zIndex": 5,
        "scale.y": {
          "0": { value: 1.03, edited: true },
          "217": { value: 1.03, edited: true, curve: "easeInOutQuad" },
          "550": { value: 0.852, edited: true }
        },
        "scale.x": {
          "0": { value: 1.025, edited: true },
          "217": { value: 1.025, edited: true, curve: "easeInOutQuad" },
          "550": { value: 0.844, edited: true }
        },
        opacity: {
          "0": { value: 0, edited: true },
          "217": { value: 0, edited: true, curve: "easeInOutQuad" },
          "550": { value: 1, edited: true }
        }
      },
      "haiku:Layer-2-97ee2c2b9a31912c": { "data-name": "Layer 2" },
      "haiku:Layer-1-2-a40a60bbda8dad28": { "data-name": "Layer 1" },
      "haiku:Ellipse-3-Copy-3-Image-a4cf88814d7eaab2": {
        "data-name": "Ellipse 3 copy 3 Image",
        width: 585,
        height: 585,
        "xlink:href": "web+haikuroot://assets/designs/SVG/Asset 1_image_1.png",
        "scale.x": 0.24,
        "scale.y": 0.24
      },
      "haiku:Ellipse-3-Copy-4-Image-5bbd5c312489e146": {
        "data-name": "Ellipse 3 copy 4 Image",
        width: 749,
        height: 839,
        "xlink:href": "web+haikuroot://assets/designs/SVG/Asset 1_image_2.png",
        "translation.x": 325,
        "translation.y": 234,
        "scale.x": 0.24,
        "scale.y": 0.24
      },
      "haiku:Blob4-a521d464cd34a306": {
        viewBox: "0 0 288.72 288.72",
        "style.position": "absolute",
        "style.margin": "0",
        "style.padding": "0",
        "style.border": "0",
        "sizeMode.x": 1,
        "sizeMode.y": 1,
        "sizeAbsolute.x": 288.72,
        "sizeAbsolute.y": 288.72,
        "translation.x": {
          "0": { value: 491.394, edited: true },
          "367": { value: 491.394, edited: true, curve: "easeInOutQuad" },
          "750": { value: 502.25237213654964, edited: true }
        },
        "translation.y": {
          "0": { value: 271.401, edited: true },
          "367": { value: 271.401, edited: true, curve: "easeInOutQuad" },
          "750": { value: 243.88191954600796, edited: true }
        },
        "style.zIndex": 6,
        "scale.x": {
          "0": { value: 1.204, edited: true },
          "367": { value: 1.204, edited: true, curve: "easeInOutQuad" },
          "750": { value: 1.069, edited: true }
        },
        "scale.y": {
          "0": { value: 1.294, edited: true },
          "367": { value: 1.294, edited: true, curve: "easeInOutQuad" },
          "750": { value: 1.147, edited: true }
        },
        opacity: {
          "0": { value: 0, edited: true },
          "367": { value: 0, edited: true, curve: "easeInOutQuad" },
          "750": { value: 1, edited: true }
        }
      },
      "haiku:Layer-2-9a206221cbf12b71": { "data-name": "Layer 2" },
      "haiku:Ellipse-3-Copy-Image-de3fbf8cb84bb46e": {
        "data-name": "Ellipse 3 copy Image"
      },
      "haiku:Ellipse-3-Copy-Image-2-58b371fad0d7f578": {
        "data-name": "Ellipse 3 copy Image",
        width: 1203,
        height: 1203,
        "xlink:href": "web+haikuroot://assets/designs/blob4_image_1.png",
        "scale.x": 0.24,
        "scale.y": 0.24
      },
      "haiku:Blob5-8a30a36d141346a7": {
        viewBox: "0 0 255.6 68.16",
        "style.position": "absolute",
        "style.margin": "0",
        "style.padding": "0",
        "style.border": "0",
        "sizeMode.x": 1,
        "sizeMode.y": 1,
        "sizeAbsolute.x": 255.6,
        "sizeAbsolute.y": 68.16,
        "translation.x": {
          "0": { value: 191.771, edited: true },
          "500": { value: 191.771, edited: true, curve: "easeInOutQuad" },
          "800": { value: 179.83354492187505, edited: true }
        },
        "translation.y": {
          "0": { value: 456.795, edited: true },
          "500": { value: 456.795, edited: true, curve: "easeInOutQuad" },
          "800": { value: 417.7084826660156, edited: true }
        },
        "style.zIndex": 7,
        opacity: {
          "0": { value: 0, edited: true },
          "500": { value: 0, edited: true, curve: "easeInOutQuad" },
          "800": { value: 1, edited: true }
        }
      },
      "haiku:Layer-2-0abdabb796c406aa": { "data-name": "Layer 2" },
      "haiku:Ellipse-3-Copy-3-Image-2d8b7bc9f85d4d69": {
        "data-name": "Ellipse 3 copy 3 Image"
      },
      "haiku:Ellipse-3-Copy-3-Image-2-2652bb6b2ab67f13": {
        "data-name": "Ellipse 3 copy 3 Image",
        width: 1065,
        height: 284,
        "xlink:href": "web+haikuroot://assets/designs/blob5_image_1.png",
        "scale.x": 0.24,
        "scale.y": 0.24
      },
      "haiku:Blob6-4c189e643cfd0281": {
        viewBox: "0 0 258.24 258.24",
        "style.position": "absolute",
        "style.margin": "0",
        "style.padding": "0",
        "style.border": "0",
        "sizeMode.x": 1,
        "sizeMode.y": 1,
        "sizeAbsolute.x": 258.24,
        "sizeAbsolute.y": 258.24,
        "translation.x": {
          "0": { value: 267.18030989173684, edited: true },
          "333": {
            value: 267.18030989173684,
            edited: true,
            curve: "easeInOutQuad"
          },
          "833": { value: 357.134, edited: true }
        },
        "translation.y": {
          "0": { value: 280.63067138671875, edited: true },
          "333": {
            value: 280.63067138671875,
            edited: true,
            curve: "easeInOutQuad"
          },
          "833": { value: 258.709, edited: true }
        },
        "style.zIndex": 8,
        opacity: {
          "0": { value: 0, edited: true },
          "333": { value: 0, edited: true, curve: "easeInOutQuad" },
          "833": { value: 1, edited: true }
        },
        "scale.x": {
          "0": { value: 1, edited: true, curve: "easeInOutQuad" },
          "500": { value: 2, edited: true, curve: "easeInOutQuad" },
          "833": { value: 1, edited: true }
        },
        "scale.y": {
          "0": { value: 1, edited: true, curve: "easeInOutQuad" },
          "500": { value: 2, edited: true, curve: "easeInOutQuad" },
          "833": { value: 1, edited: true }
        }
      },
      "haiku:Layer-2-6709a01f40a0b8b0": { "data-name": "Layer 2" },
      "haiku:Blob1-Image-d16a04ae24a75353": { "data-name": "Blob1 Image" },
      "haiku:Blob1-Image-2-d3bb889d67579bd6": {
        "data-name": "Blob1 Image",
        width: 1076,
        height: 1076,
        "xlink:href": "web+haikuroot://assets/designs/blob6_image_1.png",
        "scale.x": 0.24,
        "scale.y": 0.24
      }
    }
  },
  template: {
    elementName: "div",
    attributes: { "haiku-id": "Main-03757d2ca1026e0a", "haiku-title": "Main" },
    children: [
      {
        elementName: "svg",
        attributes: {
          version: "1.1",
          xmlns: "http://www.w3.org/2000/svg",
          "xmlns:xlink": "http://www.w3.org/1999/xlink",
          "haiku-id": "Blob6-4c189e643cfd0281",
          "haiku-title": "blob6",
          "haiku-source": "designs/blob6.svg"
        },
        children: [
          {
            elementName: "g",
            attributes: {
              "haiku-id": "Layer-2-6709a01f40a0b8b0",
              id: "Layer_2"
            },
            children: [
              {
                elementName: "g",
                attributes: {
                  "haiku-id": "Shapes-7b6e9a2471852a04",
                  id: "shapes"
                },
                children: [
                  {
                    elementName: "g",
                    attributes: {
                      "haiku-id": "Blob1-Image-d16a04ae24a75353",
                      id: "Blob1_Image"
                    },
                    children: [
                      {
                        elementName: "image",
                        attributes: {
                          "haiku-id": "Blob1-Image-2-d3bb889d67579bd6",
                          id: "Blob1_Image-2"
                        },
                        children: []
                      }
                    ]
                  }
                ]
              }
            ]
          }
        ]
      },
      {
        elementName: "svg",
        attributes: {
          version: "1.1",
          xmlns: "http://www.w3.org/2000/svg",
          "xmlns:xlink": "http://www.w3.org/1999/xlink",
          "haiku-id": "Blob5-8a30a36d141346a7",
          "haiku-title": "blob5",
          "haiku-source": "designs/blob5.svg"
        },
        children: [
          {
            elementName: "g",
            attributes: {
              "haiku-id": "Layer-2-0abdabb796c406aa",
              id: "Layer_2"
            },
            children: [
              {
                elementName: "g",
                attributes: {
                  "haiku-id": "Shapes-f32ba3c749f80258",
                  id: "shapes"
                },
                children: [
                  {
                    elementName: "g",
                    attributes: {
                      "haiku-id": "Ellipse-3-Copy-3-Image-2d8b7bc9f85d4d69",
                      id: "Ellipse_3_copy_3_Image"
                    },
                    children: [
                      {
                        elementName: "image",
                        attributes: {
                          "haiku-id":
                            "Ellipse-3-Copy-3-Image-2-2652bb6b2ab67f13",
                          id: "Ellipse_3_copy_3_Image-2"
                        },
                        children: []
                      }
                    ]
                  }
                ]
              }
            ]
          }
        ]
      },
      {
        elementName: "svg",
        attributes: {
          version: "1.1",
          xmlns: "http://www.w3.org/2000/svg",
          "xmlns:xlink": "http://www.w3.org/1999/xlink",
          "haiku-id": "Blob4-a521d464cd34a306",
          "haiku-title": "blob4",
          "haiku-source": "designs/blob4.svg"
        },
        children: [
          {
            elementName: "g",
            attributes: {
              "haiku-id": "Layer-2-9a206221cbf12b71",
              id: "Layer_2"
            },
            children: [
              {
                elementName: "g",
                attributes: {
                  "haiku-id": "Shapes-39a688fa700f7ab6",
                  id: "shapes"
                },
                children: [
                  {
                    elementName: "g",
                    attributes: {
                      "haiku-id": "Ellipse-3-Copy-Image-de3fbf8cb84bb46e",
                      id: "Ellipse_3_copy_Image"
                    },
                    children: [
                      {
                        elementName: "image",
                        attributes: {
                          "haiku-id": "Ellipse-3-Copy-Image-2-58b371fad0d7f578",
                          id: "Ellipse_3_copy_Image-2"
                        },
                        children: []
                      }
                    ]
                  }
                ]
              }
            ]
          }
        ]
      },
      {
        elementName: "svg",
        attributes: {
          version: "1.1",
          xmlns: "http://www.w3.org/2000/svg",
          "xmlns:xlink": "http://www.w3.org/1999/xlink",
          "haiku-id": "Asset-1-3dcb3812a3f867d9",
          "haiku-title": "blob34",
          "haiku-source": "designs/SVG/Asset 1.svg"
        },
        children: [
          {
            elementName: "g",
            attributes: {
              "haiku-id": "Layer-2-97ee2c2b9a31912c",
              id: "Layer_2"
            },
            children: [
              {
                elementName: "g",
                attributes: {
                  "haiku-id": "Layer-1-2-a40a60bbda8dad28",
                  id: "Layer_1-2"
                },
                children: [
                  {
                    elementName: "image",
                    attributes: {
                      "haiku-id": "Ellipse-3-Copy-3-Image-a4cf88814d7eaab2",
                      id: "Ellipse_3_copy_3_Image"
                    },
                    children: []
                  },
                  {
                    elementName: "image",
                    attributes: {
                      "haiku-id": "Ellipse-3-Copy-4-Image-5bbd5c312489e146",
                      id: "Ellipse_3_copy_4_Image"
                    },
                    children: []
                  }
                ]
              }
            ]
          }
        ]
      },
      {
        elementName: "svg",
        attributes: {
          version: "1.1",
          xmlns: "http://www.w3.org/2000/svg",
          "xmlns:xlink": "http://www.w3.org/1999/xlink",
          "haiku-id": "Layer-1-Artish2-03-f996e91f0f8c2e56",
          "haiku-title": "blob1",
          "haiku-source":
            "designs/artish2.ai.contents/artboards/artish2-03.svg",
          id: "Layer_1"
        },
        children: [
          {
            elementName: "image",
            attributes: {
              "haiku-id": "Ellipse-3-Copy-2-X-A0-Image-cdae8131d18a4358",
              id: "Ellipse_3_copy_2_xA0_Image"
            },
            children: []
          }
        ]
      },
      {
        elementName: "svg",
        attributes: {
          version: "1.1",
          xmlns: "http://www.w3.org/2000/svg",
          "xmlns:xlink": "http://www.w3.org/1999/xlink",
          "haiku-id": "Layer-1-Artish2-04-d1c1411a17f9353a",
          "haiku-title": "wave1",
          "haiku-source":
            "designs/artish2.ai.contents/artboards/artish2-04.svg",
          id: "Layer_1"
        },
        children: [
          {
            elementName: "image",
            attributes: {
              "haiku-id": "Ellipse-2-Copy-3-X-A0-Image-4ecacd5172c17398",
              id: "Ellipse_2_copy_3_xA0_Image"
            },
            children: []
          }
        ]
      },
      {
        elementName: "svg",
        attributes: {
          version: "1.1",
          xmlns: "http://www.w3.org/2000/svg",
          "xmlns:xlink": "http://www.w3.org/1999/xlink",
          "haiku-id": "Layer-1-Artish2-07-4f9f69f6a48787db",
          "haiku-title": "wave2",
          "haiku-source":
            "designs/artish2.ai.contents/artboards/artish2-07.svg",
          id: "Layer_1"
        },
        children: [
          {
            elementName: "image",
            attributes: {
              "haiku-id": "Ellipse-2-Copy-2-X-A0-Image-667e160c068a8e09",
              id: "Ellipse_2_copy_2_xA0_Image"
            },
            children: []
          }
        ]
      },
      {
        elementName: "svg",
        attributes: {
          version: "1.1",
          xmlns: "http://www.w3.org/2000/svg",
          "xmlns:xlink": "http://www.w3.org/1999/xlink",
          "haiku-id": "Layer-1-Artish2-01-85cec093356d01e8",
          "haiku-title": "background",
          "haiku-source":
            "designs/artish2.ai.contents/artboards/artish2-01.svg",
          id: "Layer_1"
        },
        children: [
          {
            elementName: "image",
            attributes: {
              "haiku-id": "Rectangle-1-X-A0-Image-c6751370fcdb39b2",
              id: "Rectangle_1_xA0_Image"
            },
            children: []
          }
        ]
      }
    ]
  }
};
