// @haiku/core/lib/adapters/vue-dom/HaikuVueDOMAdapter.js
"use strict";
/* tslint:disable:max-line-length */
/**
 * Copyright (c) Haiku 2016-2018. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var getParsedProperty_1 = require("../../helpers/getParsedProperty");
var StringUtils_1 = require("../../helpers/StringUtils");
var clearProps = function (props) {
    var result = {};
    for (var verboseKeyName in props) {
        if (props[verboseKeyName] === undefined) {
            continue;
        }
        Object.assign(result, getParsedProperty_1.default(props, verboseKeyName));
    }
    return result;
};
var allProps = function (vueComponent) {
    return Object.assign(clearProps(vueComponent.$props), {
        ref: vueComponent.$el,
        onHaikuComponentWillInitialize: function (component) {
            vueComponent.$emit('haikuComponentWillInitialize', component);
        },
        onHaikuComponentDidMount: function (component) {
            vueComponent.$emit('haikuComponentDidMount', component);
        },
        onHaikuComponentWillMount: function (component) {
            vueComponent.$emit('haikuComponentWillMount', component);
        },
        onHaikuComponentDidInitialize: function (component) {
            vueComponent.$emit('haikuComponentDidInitialize', component);
        },
        onHaikuComponentWillUnmount: function (component) {
            vueComponent.$emit('haikuComponentWillUnmount', component);
        },
        children: vueComponent.$slots.default
            ? vueComponent.$slots.default.filter(function (node) { return node.tag !== undefined; })
            : [],
        vanities: {
            'controlFlow.placeholder': function (element, surrogate, value, context, timeline, receiver, sender) {
                if (element.__memory.placeholder.surrogate === surrogate || !element.__memory.targets) {
                    return;
                }
                var node = element.__memory.targets[0];
                if (node) {
                    var vueElement = surrogate.elm;
                    var div = document.createElement('div');
                    node.parentNode.replaceChild(div, node);
                    node.style.visibility = 'hidden';
                    if (vueElement) {
                        div.appendChild(vueElement);
                    }
                    window.requestAnimationFrame(function () {
                        element.__memory.placeholder.surrogate = surrogate;
                        node.style.visibility = 'visible';
                    });
                    sender.markHorizonElement(element);
                    sender.markForFullFlush();
                }
            },
        },
    });
};
// tslint:disable-next-line:function-name
function HaikuVueDOMAdapter(haikuComponentFactory) {
    return {
        props: {
            // We use null (which is the equivalent of 'any') for Boolean values
            // because Vue does typecasting for us, which sets undefined to false.
            automount: null,
            autoplay: null,
            forceFlush: null,
            freeze: null,
            loop: null,
            alwaysComputeSizing: null,
            seed: String,
            sizing: String,
            timestamp: Number,
            frame: Function,
            clock: Object,
            preserve3d: String,
            contextMenu: String,
            position: String,
            overflowX: String,
            overflowY: String,
            overflow: String,
            mixpanel: String,
            interactionMode: Object,
            states: Object,
            eventHandlers: Object,
            timelines: Object,
            vanities: Object,
            children: Array,
            placeholder: Object,
            // LEGACY
            haikuOptions: Object,
        },
        mounted: function () {
            this.haiku = haikuComponentFactory(this.$el, allProps(this));
        },
        updated: function () {
            this.haiku.assignConfig(allProps(this));
        },
        destroyed: function () {
            this.haiku.callUnmount();
        },
        render: function (createElement) {
            return createElement('div', {
                attrs: {
                    id: "haiku-vueroot-" + StringUtils_1.randomString(24),
                },
                style: {
                    position: 'relative',
                    margin: 0,
                    padding: 0,
                    border: 0,
                    width: '100%',
                    height: '100%',
                    transform: 'matrix3d(1,0,0,0,0,1,0,0,0,0,1,0,0,0,0,1)',
                },
            }, this.$slots.default);
        },
    };
}
exports.default = HaikuVueDOMAdapter;
